nomad/node_endpoint.go

```diff
@@ -1313,6 +1339,17 @@ func (n *Node) UpdateAlloc(args *structs.AllocUpdateRequest, reply *structs.Gene
 		return err
 	}
 
+	// Update node alloc update index.
+	copyNode := node.Copy()
+	copyNode.LastAllocUpdateIndex = future.Index()
+
+	_, _, err = n.srv.raftApply(structs.NodeRegisterRequestType, &structs.NodeRegisterRequest{
+		Node: copyNode,
+	})
+	if err != nil {
+		return fmt.Errorf("node update failed: %v", err)
+	}
```

Comment:
With this change there are two raft logs in this RPC. It's unlikely, but if there's a failure between the batched alloc/eval update and this one, we'll end up with the `LastAllocUpdateIndex` not having been updated. If I'm reasoning thru the logic above correctly, that would "freeze" the node in an initializing state until something there's a new allocation change on the client (which could be a while!).

It looks like the `AllocClientUpdateRequestType` raft entry is only ever used by this RPC. Could we perform the update for the node's `LastAllocUpdateIndex` in the same raft apply? 

(Another way to handle that, given that we already have to query for allocations, would be to derive the last update from the allocation's modify index. But that seems like we'd risk server-side updates to allocations causing a node to flap from ready to initializing?)

Reply:
Oh right, good catch. I think I misunderstood how the batch updates worked so I thought we wouldn't be able to reuse them. But I see that it may actually be possible.

I also thought about looking at the lowest alloc `ModifyIndex` but I figured there could be other things updating this value (like the server itself).

I will make the index update in the `AllocClientUpdateRequestType` request, thanks!

Reply:
Fixed in https://github.com/hashicorp/nomad/pull/15808/commits/952b9712178d17023dc19cde0dac007d0bf6a421.